Added linalg.tensorinv

IvanYashchuk opened the pull request:
This PR adds `torch.linalg.tensorinv` for NumPy compatibility.

Ref #42666

IvanYashchuk committed:
Added linalg_tensorinv

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8319,6 +8319,10 @@
   python_module: linalg
   variants: function
 
+- func: linalg_tensorinv(Tensor self, int? ind=2) -> Tensor
```
What's the plan for out support?

mruberry replied:
Can you elaborate on how `ind` is parsed? It's both optional and it has a default value of 2?

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1601,6 +1601,24 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
 
+Tensor linalg_tensorinv(const Tensor& self, optional<int64_t> ind) {
+  int64_t ind_value = ind.has_value() ? ind.value() : 2;
+  TORCH_CHECK(ind_value > 0, "'ind' must be a positive integer");
```
nit: "Expected a strictly positive integer for 'ind', but got " ...

mruberry replied:
Add a comment here explaining why this doesn't check that self is square

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1601,6 +1601,24 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
   return linalg_norm_out_impl(result, self, c10::nullopt, ord, opt_dim, keepdim, opt_dtype);
 }
 
+Tensor linalg_tensorinv(const Tensor& self, optional<int64_t> ind) {
```
Add a comment briefly explaining how this function computes the inverse

mruberry replied:
What, in particular, happens if self is singular?

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,64 @@ def test_nuclear_norm_exceptions_old(self, device):
 
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
```
We now require NumPy to run the test suite so we don't need checks like this anymore

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,64 @@ def test_nuclear_norm_exceptions_old(self, device):
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
```
Nit: remove the complex dtypes from @dtypes. @dtypes is generally the lowest common denominator of support, and complex functionality is very new.

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,64 @@ def test_nuclear_norm_exceptions_old(self, device):
+    def test_tensorinv(self, device, dtype):
+
+        def run_test(a_shape, ind):
+            a = torch.rand(a_shape, dtype=dtype, device=device)
```
Is there a slightly more interesting range of values? (-2, 2)?

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,64 @@ def test_nuclear_norm_exceptions_old(self, device):
+            expected = np.linalg.tensorinv(a_numpy, ind=ind)
+            self.assertEqual(result, expected)
+
+        def check_shape(a_shape, ind):
```
Separate these checks into a separate error-checking / bounds checking test. A few notes here:

- must be a single tensor or batch of tensors
- should also check that a is square
- should also check that a is non-singular

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,64 @@ def test_nuclear_norm_exceptions_old(self, device):
+        check_ind((12, 3, 4), ind=-1)
+        check_ind((18, 3, 3, 2), ind=0)
+
+    # TODO: once "solve_inverse" supports complex dtypes, they shall be added to above test
```
Organizational note: on the linalg tracker we should add tasks like "complex CUDA tensorinv" support after these PRs go in so we don't forget these issues.

IvanYashchuk replied:
In this case, complex CUDA support for `tensorinv` will come automatically when https://github.com/pytorch/pytorch/pull/45034 gets merged.
xfailing tests will remind about these issues 🙂

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,64 @@ def test_nuclear_norm_exceptions_old(self, device):
+                torch.linalg.tensorinv(a, ind=ind)
+
+        # compare to NumPy output
+        run_test((12, 3, 4), ind=1)
```
Should there be a separate test for degenerate edge cases like empty tensors and tensors with a zero size for some dims?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,28 @@
+tensorinv = _add_docstr(_linalg.linalg_tensorinv, r"""
+linalg.tensorinv(input, ind=2) -> Tensor
+
+Returns the inverse of a given tensor.
```
I find this function really confusing.

Let's say it was accurately described as "Computes the inverse of :input:`attr`, a matrix." Then I would expect this to work:

```
a = np.random.randn(4, 4)
np.linalg.tensorinv(a)
```

but instead it yells at me:
```
LinAlgError: Last 2 dimensions of the array must be square
```

Instead I think the second sentence is more accurate and should replace the first one:

"Computes a tensor inv = tensorinv(input, ind) such that tensordot(inv, ind) == I_n, where I_n is the n-dimensional identity tensor."

That sentence will also also need to define what n is.

Then I would include a discussion of ind, since it's far from intuitive.

mruberry replied:
Another idea would be to create a note describing the difference between this function and inv (currently inverse in PyTorch: https://pytorch.org/docs/master/generated/torch.inverse.html?highlight=inv#torch.inverse).

IvanYashchuk replied:
NumPy does not have any checks for the condition specified in the docs: the input must satisfy `prod(a.shape[:ind]) == prod(a.shape[ind:])`. I'll add such check here.
So your example doesn't work because `np.prod(a.shape[:2])` is 9 while `np.prod(a.shape[2:])` is 1.0, so the input is not "square" in their definition.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,28 @@
+where ``I`` denotes the identity tensor.
+
+Args:
+    input (Tensor): The tensor to invert. Its shape must satidsfy ``prod(input.shape[:ind]) == prod(input.shape[ind:])``.
```
"satidsfy" -> "satisfy"

Are there other requirements on input? Like it being invertible?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,28 @@
+Args:
+    input (Tensor): The tensor to invert. Its shape must satidsfy ``prod(input.shape[:ind]) == prod(input.shape[ind:])``.
+
+    ind (int): The positive number used in ``dims`` option of ``tensordot``.
```
This should describe how it affects the tensorinv computation, not a hypothetical tensordot computation. The conceptual aspect of how this relates to tensordot is covered in operator's description.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,28 @@
+    ind (int): The positive number used in ``dims`` option of ``tensordot``.
+
+Returns:
+    (Tensor): the inverse of ``input`` with shape ``input.shape[ind:] + input.shape[:ind]``.
```
This is probably redundant with the description and can be removed.

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,64 @@ def test_nuclear_norm_exceptions_old(self, device):
+        # compare to NumPy output
+        run_test((12, 3, 4), ind=1)
+        run_test((3, 8, 24), ind=2)
+        run_test((18, 3, 3, 2), ind=1)
```
Add some non-contiguous test cases, too. One where the tensor is transposed and another where every other element is skipped ([::2])

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,28 @@
+
+This function computes a tensor such that ``tensordot(tensorinv(input), input, ind) == I``,
+where ``I`` denotes the identity tensor.
+
```
A "Supports..." statement like that for torch.sub should appear here.

Does this support autograd?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,28 @@
     (tensor(3.7417), tensor(11.2250))
 """)
+
+tensorinv = _add_docstr(_linalg.linalg_tensorinv, r"""
```
Just to be clear, this function does not support batching, right?

It seems like it would be very difficult to make it support batching, too, with its current signature.

IvanYashchuk replied:
No, it doesn't support batching. I think we could support batching with an additional keyword argument specifying the batching dimensions, but before doing it here it should be done for `tensordot`.

mruberry commented:
Hey @IvanYashchuk! Thanks for this PR. I asked some questions and made some suggestions, mostly around the tests and docs. (as usual). 

Looking forward to an update!

IvanYashchuk committed:
Added a comment for tensorinv implementation

IvanYashchuk committed:
Modified error message for ind

IvanYashchuk committed:
Remove NumPy skips

IvanYashchuk committed:
Do not use c10::optional

IvanYashchuk committed:
Corrected test

IvanYashchuk committed:
Added check for squareness of the input tensor

Updated tests

IvanYashchuk committed:
Added a note on squareness requirement to docs

IvanYashchuk committed:
Added entry to overrides.py

IvanYashchuk committed:
Added entry to common_methods_invocations.py

IvanYashchuk committed:
Added error catch if the input is singular. Added tests

IvanYashchuk committed:
Changed rand -> randn

IvanYashchuk committed:
Added test_tensorinv_empty

IvanYashchuk committed:
Added test_tensorinv_non_contiguous

IvanYashchuk committed:
Updated documentation

IvanYashchuk committed:
Added out= support

IvanYashchuk committed:
Updated documentation.

x -> input_inv
Added a note and an example on relation to torch.inverse

IvanYashchuk commented:
@mruberry, thank you for the comments! I've updated the pull request.

- I've added `out=` support
- added error checks for 'squareness' of the input tensor and tests for that
- added catching the error from `at::inverse` in case of non-invertible input and tests for that
- added tests cases for empty, non-contiguous inputs
- autograd and jit works, tested via adding an entry to `common_methods_invocations.py`
- updated documentation

This functions is different from other linalg functions as it deals with tensors, while most of other linalg functions are for matrices and batches of matrices. Here batching is not supported, and the input tensor is one whole object that is being inverted.

IvanYashchuk committed:
Removed unused import

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-tensorinv

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,68 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  Step 2. reshape `self` to 2D matrix.
+  Step 3. invert the 2D matrix self.to_2D()
+          there is no quick way to find out whether the matrix is invertible,
+          so at this stage an error from at::inverse can be thrown
```
This is a very helpful description. How does this error checking work on CUDA? Does it cause cross-device memory movement? (However it works, a brief statement should be added here explaining it.)

IvanYashchuk replied:
So `at::inverse` calls to cusolver's `getrf` & `getrs`. Error checking is done via checking the `devInfo` integer ([cusolver docs link](https://docs.nvidia.com/cuda/cusolver/index.html#cuSolverDN-lt-t-gt-getrf)), if it is larger than 0, then there is a problem. Since it lies in the GPU memory, checking the error causes cross-device memory movement. That's an expensive thing to do. For example, CuPy has decided not to check for errors by default to avoid a performance drop.

mruberry replied:
I also think we should consider an alternative method to check for errors. cc @ngimel who's thought about this issue, too. 

As long as we mention the current behavior here (relying on at::inverse, which syncs) I think we can keep the current behavior.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,68 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+
+  // Check whether the self tensor can be reshaped to the 2D square matrix
+  TORCH_CHECK(prod_ind_end == prod_start_ind,
+    "Expected self to be a 'square' tensor such that prod(self.shape[ind:]) == prod(self.shape[:ind]), but got ",
```
This is a great check + error message.

IvanYashchuk committed:
Added a note that CUDA error checking causes cross-device memory
movement

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,69 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  TORCH_CHECK(result.sizes().equals(expected_result_shape),
+    "Expected result tensor to have size of ", expected_result_shape, ", but got tensor of size ", result.sizes());
+
+  Tensor result_tmp = at::linalg_tensorinv(self, ind);
```
Why not just compute tensorinv and then compare with the result shape?

IvanYashchuk replied:
To avoid possibly wasteful expensive computations, hoping that calculating the shapes is a comparatively cheap thing to do.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8875,6 +8875,14 @@
 
+- func: linalg_tensorinv(Tensor self, int ind=2) -> Tensor
+  python_module: linalg
+  variants: function
```
Since the recent dispatch update, this needs to add a dispatch. Since it's composed of existing PyTorch operators I think it (and its out variant) get a "math" dispatch.

IvanYashchuk replied:
The same answer as here https://github.com/pytorch/pytorch/pull/45358#discussion_r510118744

If there is no dispatch section, then "math" is assumed implicitly. Do we want it to be there explicitly?

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,171 @@ def test_nuclear_norm_exceptions_old(self, device):
+        check_singular_input((12, 3, 4), ind=1)
+        check_singular_input((3, 6, 18), ind=2)
+
+    # TODO: torch.inverse on CUDA for single matrices does not raise the error
```
This issue has been closed, so this probably needs an update.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,43 @@
+
+Computes a tensor ``input_inv`` such that ``tensordot(input_inv, input, ind) == I_n``,
+where ``I_n`` is the n-dimensional identity tensor and ``n`` is equal to ``input.ndim``.
+The resulting tensor ``input_inv`` has the shape equal to ``input.shape[ind:] + input.shape[:ind]``.
```
"has the shape equal to" -> "has shape equal to"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,43 @@
+
+Supports real and, only on the CPU, complex inputs.
+
+.. note:: If :attr:`input` is not a 'square' tensor, meaning it does not satisfy the requirement
```
I would remove the idea of input being "square" since "square" = nxn matrix. I think this can just state the requirement plainly. 

"If :attr:\`input\` is not invertible or \`\`prod(input.shape[ind:]) != prod(input.shape[:ind])\`\` then a RuntimeError is thrown." 

Note that the first part of this sentence is only recently true on CUDA.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,43 @@
+          ``prod(input.shape[ind:]) == prod(input.shape[:ind])`` or if :attr:`input` is not invertible,
+          then a RuntimeError will be thrown.
+
+.. note:: If :attr:`input` is a 2-dimensional tensor and ``ind=1``, then this function computes matrix inverse.
```
"If" -> "When" 

", then this function compute matrix inverse" -> "this function computes the (multiplicative) inverse of :attr:\`input\`, equivalent to calling :func:\`torch.inverse\`."

(That sample func reference may need a formatting fix.)

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,43 @@
+.. note:: If :attr:`input` is a 2-dimensional tensor and ``ind=1``, then this function computes matrix inverse.
+
+Args:
+    input (Tensor): The tensor to invert. Its shape must satisfy ``prod(input.shape[:ind]) == prod(input.shape[ind:])``.
```
"The tensor" -> "A tensor"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,43 @@
+
+Args:
+    input (Tensor): The tensor to invert. Its shape must satisfy ``prod(input.shape[:ind]) == prod(input.shape[ind:])``.
+    ind (int): The positive integer representing how many first indices are involved in the inverse contraction.
```
"The positive" -> "A positive"

IvanYashchuk replied:
I agree that it's better to avoid "squareness", not to confuse people. What is the inverse equation, then?
I think I'll change it to "A positive integer that affects the result's shape.  Default: 2." And how it affects the shape is described in the main paragraph.

mruberry replied:
That's OK. You're right that my own suggestion was a little vague. By "inverse equation" I meant the tensordot expression in the op's description. Maybe we can reference that? I like your suggestion but also find "affects the result's shape" to be a little vague.

IvanYashchuk committed:
Remove mentioning 'square' tensor term

IvanYashchuk committed:
Updated documentation

Changed ind argument description
Removed "square" term

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-tensorinv

IvanYashchuk committed:
Enable test for singular input on CUDA

IvanYashchuk committed:
Removed unused import

heitorschueroff commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,69 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  // self[ind:] + self[:ind]
+  expected_result_shape.insert(expected_result_shape.end(), shape_start_ind.begin(), shape_start_ind.end());
+
+  TORCH_CHECK(result.sizes().equals(expected_result_shape),
```
Use at::native::resize_output in native/Resize.cpp file to check and resize output. We should allow out tensors to be empty in which case we resize them.

IvanYashchuk replied:
Thanks for the suggestion! I've added it now.

mruberry replied:
(Reminder for both of us: same issue applies to other PRs, too.)

heitorschueroff commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,69 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+    "Expected result tensor to have size of ", expected_result_shape, ", but got tensor of size ", result.sizes());
+
+  Tensor result_tmp = at::linalg_tensorinv(self, ind);
+  result.copy_(result_tmp);
```
at::inverse has an out version, can we use that to avoid the extra copy? Maybe reduce at::linalg_tensorinv without out to the one with out by passing an empty tensor, with my above comment, that should work. Only thing to be careful is that out variants do not compute gradients.

mruberry replied:
In this case I don't think the out version of the op can be called because this is a composite op without its own autograd entry.

IvanYashchuk replied:
Okay, I've implemented that `linalg_tensorinv_out` is primary and `linalg_tensorinv` calls to it, but autograd does not work in this case.
```
test/test_autograd.py::TestAutogradDeviceTypeCPU::test_linalg.tensorinv_cpu inverse(): functions with out=... arguments don't support automatic differentiation, but one of the arguments requires grad.
```

mruberry replied:
Correct. You can't use this approach. You need to keep linalg_tensorinv calling only differentiable ops (no out variants there).

heitorschueroff replied:
If you want to pursue this opt you could create a static helper function that takes an optional out parameter, this way you can share most of the code but add logic that if it has an out parameter you call at::inverse_out.

ngimel replied:
linalg functions are so slow that this microoptimization of avoiding copy at the end is really not worth code complexity.

heitorschueroff commented on test/test_linalg.py:
```diff
@@ -921,6 +922,156 @@ def test_nuclear_norm_exceptions_old(self, device):
+    @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
+    @dtypesIfCUDA(torch.float32, torch.float64)
+    @precisionOverride({torch.float: 1e-3, torch.cfloat: 1e-3})
+    def test_tensorinv_non_contiguous(self, device, dtype):
```
Add a test for the out parameter where the given out tensor is discontiguous as well, it should not change the sizes and strides of the provided out tensor.

IvanYashchuk replied:
Done.

IvanYashchuk committed:
Added non-contiguous out test case

IvanYashchuk committed:
Make _out function implementation primary

IvanYashchuk committed:
Make non out function primary back, fixes autograd

IvanYashchuk committed:
Updated errors and warnings tests

IvanYashchuk committed:
Make sure that the input is invertible in tests

IvanYashchuk committed:
Added dispatch: Math

IvanYashchuk committed:
Skip annotation testing with exclude_tensor_method

IvanYashchuk committed:
Moved non-contiguous out test case

IvanYashchuk committed:
flake8

glaringlee commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,57 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  Tensor result;
+  try {
+    result = at::inverse(self.reshape({prod_ind_end, prod_ind_end}));
+  } catch (...) {
```
Just curious on this catch block. In C++, we normally don't prefer catch (...) if we explicitly know the exception type (which is `c10::Error` in this case). I guess using (...) here is because at::inverse throw exception only when the input is singular, correct?

IvanYashchuk replied:
Yes, at::inverse throws exception only when the input is not a square matrix (that's checked here before the call) and when the input is singular.
I can change that to use the explicit exception type.

glaringlee replied:
I see. I think we can just keep (...) in this case then.

ezyang replied:
I disagree. See https://github.com/pytorch/pytorch/issues/53739

IvanYashchuk committed:
Changed 'ind' description

IvanYashchuk commented:
A note: once #45034 gets merged, this PR needs to be updated.

- [ ] Remove `test_tensorinv_xfailed` and add complex dtypes to main tests
- [ ] Update [in `common_methods_invocations.py`](https://github.com/pytorch/pytorch/pull/45969/files#diff-18a0c3e7cd07089b34760028c387d39de1aa393172f1d5f123692106b3a3d144R1424) lambda function arguments.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-tensorinv

IvanYashchuk committed:
Merge branch 'master' into numpy-tensorinv

mruberry commented:
Hey @IvanYashchuk, just checking in on this PR, too. How's it going?

IvanYashchuk commented:
> A note: once #45034 gets merged, this PR needs to be updated.
> 
> * [ ]  Remove `test_tensorinv_xfailed` and add complex dtypes to main tests
> * [ ]  Update [in `common_methods_invocations.py`](https://github.com/pytorch/pytorch/pull/45969/files#diff-18a0c3e7cd07089b34760028c387d39de1aa393172f1d5f123692106b3a3d144R1424) lambda function arguments.

#45034 was merged, I'll update this PR and it should be ready for merging after that.

IvanYashchuk commented:
#45034 was reverted, waiting now for https://github.com/pytorch/pytorch/pull/47595 to be merged.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-tensorinv

IvanYashchuk committed:
tensorinv complex now works on CUDA

IvanYashchuk commented:
`torch.inverse` for complex on CUDA was merged and I've updated this PR.

@mruberry, this is ready for another (last?) review round.

Updated `linalg_tensorinv_out` to use `at::native::resize_output`.
Added non-contiguous `out=` tests cases.
Updated documentation: changed `ind` argument description; removed "square tensor" term.

IvanYashchuk committed:
Merge branch 'master' into numpy-tensorinv

IvanYashchuk committed:
Merge branch 'master' into numpy-tensorinv

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1604,6 +1604,57 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  // self[:ind]
+  std::vector<int64_t> shape_start_ind = self.sizes().slice(0, ind).vec();
+
+  int64_t prod_ind_end = std::accumulate(shape_ind_end.begin(), shape_ind_end.end(), int64_t{1}, std::multiplies<int64_t>());
```
nit: cbegin/cend where possible

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1604,6 +1604,57 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+  return result.reshape(shape_ind_end);
+}
+
+Tensor& linalg_tensorinv_out(Tensor& result, const Tensor& self, int64_t ind) {
```
Add a TODO comment here that we should reuse result's storage instead of having to allocate another output and copy to result

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -214,6 +214,47 @@
+where ``I_n`` is the n-dimensional identity tensor and ``n`` is equal to ``input.ndim``.
+The resulting tensor ``input_inv`` has shape equal to ``input.shape[ind:] + input.shape[:ind]``.
+
+Supports real-valued and complex-valued inputs.
```
"real-valued" -> "float, double,"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -214,6 +214,47 @@
+Args:
+    input (Tensor): A tensor to invert. Its shape must satisfy ``prod(input.shape[:ind]) == prod(input.shape[ind:])``.
+    ind (int): A positive integer that describes the inverse tensor equation. See :func:`torch.tensordot` for details.
+               Default: 2.
```
No newline here.

mruberry reviewed (approved):
Nice work, @IvanYashchuk!

Made a couple small suggestions. Just ping me when you'd like this merged.

IvanYashchuk committed:
Use const_iterators

IvanYashchuk committed:
Added a todo comment

IvanYashchuk committed:
List explicitly supported dtypes

IvanYashchuk committed:
Doc fix: no newline for 'ind' description

IvanYashchuk commented:
Thank you, @mruberry! This should be ready for merge now.

mruberry commented:
Unfortunately the internal tools still fail to understand how to work with method_tests() properly, so as usual we'll need to remove them for now. As I always say, in the future we'll use OpInfo-based tests for these instead.

Would you rip out the common_methods_invocations.py changes and ping me, @IvanYashchuk? Sorry to make you do it.

IvanYashchuk committed:
Removed changes to common_methods_invocations.py

IvanYashchuk commented:
@mruberry I removed those changes.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-tensorinv
